packages/eslint-plugin/src/rules/prefer-reduce-type-parameter.ts

```diff
@@ -0,0 +1,116 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+import * as ts from 'typescript';
+import * as util from '../util';
+
+const getMemberExpressionName = (
+  member: TSESTree.MemberExpression | TSESTree.OptionalMemberExpression,
+): string | null => {
+  if (!member.computed) {
+    return member.property.name;
+  }
+
+  if (
+    member.property.type === AST_NODE_TYPES.Literal &&
+    typeof member.property.value === 'string'
+  ) {
+    return member.property.value;
+  }
+
+  return null;
+};
+
+export default util.createRule({
+  name: 'prefer-reduce-type-parameter',
+  meta: {
+    type: 'problem',
+    docs: {
+      category: 'Best Practices',
+      recommended: 'error',
+      description:
+        'Prefer using type parameter when calling `Array#reduce` instead of casting',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferTypeParameter:
+        'Unnecessary cast: Array#reduce accepts a type parameter for the default value.',
+    },
+    fixable: 'code',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const service = util.getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    const reportPossibleArrayReduceCasting = (
+      node: TSESTree.CallExpression | TSESTree.OptionalCallExpression,
+    ): void => {
+      {
+        if (
+          !util.isMemberOrOptionalMemberExpression(node.callee) ||
+          getMemberExpressionName(node.callee) !== 'reduce'
+        ) {
+          return;
+        }
+
+        const [, secondArg] = node.arguments;
+
+        if (
+          node.arguments.length < 2 ||
+          (secondArg.type !== AST_NODE_TYPES.TSAsExpression &&
+            secondArg.type !== AST_NODE_TYPES.TSTypeAssertion)
+        ) {
+          return;
+        }
+
+        // Get the symbol of the `reduce` method.
+        const tsNode = service.esTreeNodeToTSNodeMap.get(node.callee);
+        const reduceSymbol = checker.getSymbolAtLocation(tsNode);
+        if (reduceSymbol == null) {
+          return;
+        }
+
+        // Check the owner type of the `reduce` method.
+        for (const methodDecl of reduceSymbol.declarations) {
+          const typeDecl = methodDecl.parent;
+          if (
+            ts.isInterfaceDeclaration(typeDecl) &&
+            ts.isSourceFile(typeDecl.parent) &&
+            typeDecl.name.escapedText === 'Array'
+          ) {
```

Comment:
note - you can use these utility functions to check if the object on the member expression is an array.

https://github.com/typescript-eslint/typescript-eslint/blob/bc2a9d6184973f1076e9aa4404fec3207b0107d7/packages/eslint-plugin/typings/typescript.d.ts#L7-L20

Reply:
Awesome! That came from `require-sort-comparator` rule, so that can be refactored to use this utils too.

Happy to apply this to this PR now, or can do a sweep of all the rules in a follow up PR in one commit :)

Reply:
I would prefer this rule goes out with the correct code, and we patch existing code later.

Reply:
I figured out why that utility isn't used in `require-array-sort-compare`: it doesn't seem to work on generics that extend an array:

```
  ● prefer-reduce-type-parameter › invalid ›
function f<T, U extends T[]>(a: U) {
  return a.reduce(() => {}, {} as Record<string, boolean>);
}


    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      1
    Received:
      0

    Message:
      Should have 1 error but had 0: []

      at testInvalidTemplate (../../node_modules/eslint/lib/rule-tester/rule-tester.js:507:24)
      at Object.<anonymous> (../../node_modules/eslint/lib/rule-tester/rule-tester.js:668:25)
```

Reply:
oh really? That kind of makes sense. I believe that you have to use our `getConstrainedTypeAtLocation` utility function to adjust the type appropriately.

I think that this is something a lot of our rules are doing wrong...

Reply:
> I believe that you have to use our getConstrainedTypeAtLocation utility function

Yeah that does the trick - awesome.

> I think that this is something a lot of our rules are doing wrong...

I assume you mean they're not using `getConstrainedTypeAtLocation`.
I'm happy to look into this if you'd like; feel free to make an issue & assign it to me.

Reply:
Ugh, turns out that `getConstrainedTypeAtLocation` doesn't work for unions :/

```
  ● prefer-reduce-type-parameter › invalid ›
function f(a: string[] \u007c number[]) {
  return a.reduce(() => {}, {} as Record<string, boolean>);
}


    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      1
    Received:
      0

    Message:
      Should have 1 error but had 0: []

      at testInvalidTemplate (../../node_modules/eslint/lib/rule-tester/rule-tester.js:507:24)
      at Object.<anonymous> (../../node_modules/eslint/lib/rule-tester/rule-tester.js:668:25)
```

(Also for some reason something is causing the pipe to be rendered as `\u007c` for some reason 😕)

Reply:
> Also for some reason something is causing the pipe to be rendered as \u007c in my console

This is a bug in ESLint (https://github.com/eslint/eslint/pull/12771). Unfortunately I fixed that bug _after_ they had started working on the v7 release, so it won't be out until then 😢 

----

> doesn't work for unions

It wouldn't, no, because a union isn't an array type!

However, it's worth noting that this code also fails the type checker. Having a quick play it also looks like there's no valid union of arrays which passes type checking for a call to `reduce`.
So I don't think it's a problem if we "false negative" in this case, because the code wouldn't even compile for the user.

Though note that it's pretty trivial to do with the help of the `unionTypeParts` utility function from `tsutils`:

```ts
import { unionTypeParts } from 'tsutils';

function isTypeArrayTypeOrUnionOfArrayTypes(
  type: ts.Type,
  ts.TypeChecker,
): boolean {
  for (const t of unionTypeParts(type)) {
    if (!checker.isArrayType(t)) {
      return false;
    }
  }

  return true;
}
```

Reply:
> looks like there's no valid union of arrays which passes type checking for a call to reduce.

Ah yes interesting - but that's not the case for `sort` 😉

I've made this util & used it in #1728